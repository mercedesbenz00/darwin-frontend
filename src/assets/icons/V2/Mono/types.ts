export enum IconVariant {
  DEFAULT = 'default',
  DISABLED = 'disabled',
  SELECTED = 'selected',
  NEGATIVE = 'negative',
}
