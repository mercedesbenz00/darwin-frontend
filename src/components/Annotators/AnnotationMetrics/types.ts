import { MembershipPayload, DatasetAnnotationReportPayload } from '@/store/types'

export type AnnotationDataRange = 'total' | 'month' | 'week' | 'day'
export type AnnotationDataGranularity = 'day' | 'hour'

export type ChartDimension =
  'annotationTime' |
  'imagesAnnotated' |
  'avgTimePerAnnotation' |
  'annotationsCreated' |
  'reviewPassRate'

export type AdaptedDataPoint =
  {[k in Exclude<ChartDimension, 'reviewPassRate' | 'avgTimePerAnnotation'>]: number } &
  {[k in Extract<ChartDimension, 'reviewPassRate' | 'avgTimePerAnnotation'>]: number | null} &
  { userId: number | null, date: string }

export type AggregatedDataPoint = Pick<
    DatasetAnnotationReportPayload,
    | 'annotation_time'
    | 'annotations_approved'
    | 'annotations_created'
    | 'date'
    | 'images_annotated'
    | 'images_approved'
    | 'images_rejected'
    | 'user_id'
  >

export type AnnotationDimensionData = {
  annotationTime: number,
  annotationsCreated: number,
  avgTimePerAnnotation: number | null,
  imagesAnnotated: number,
  reviewPassRate: number | null
}

export type AnnotationChartData = {
  data: AdaptedDataPoint[]
  dimensions: AnnotationDimensionData
  member?: MembershipPayload
  visible: boolean
}

export type AnnotationBarChartData = {
  data: AggregatedDataPoint[]
  member?: MembershipPayload
}

/**
 * Format for a single item to be shown in a chart tooltip.
 *
 * A tooltip will contain multiple items
 */
export type ChartTooltipData = { member: MembershipPayload | null, value: string}
