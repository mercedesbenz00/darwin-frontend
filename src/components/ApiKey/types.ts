export type AbilityGroup =
  'manage_comments' |
  'manage_tasks' |
  'manage_datasets' |
  'manage_team' |
  'manage_team_members' |
  'manage_models'

export type Group = { id: AbilityGroup; name: string }
export type GroupOption = Group & { selected: boolean }

export enum SupportedLanguages {
  shell = 'shell',
  python = 'python',
  javascript = 'javascript',
  elixir = 'elixir',
  cli = 'cli'
}

export type AccessCodeExample = { language: SupportedLanguages, code: string }
