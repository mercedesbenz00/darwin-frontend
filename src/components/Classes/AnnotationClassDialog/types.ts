import { AnnotationClassPayload, AnnotationClassMetadata, AnnotationTypeName } from '@/store/types'

type PendingFields =
  | 'crop_key'
  | 'crop_url'
  | 'image_height'
  | 'image_width'
  | 'key'
  | 'original_image_url'
  | 'scale'
  | 'x'
  | 'y'

/**
 * Holds a class image that isn't necessarily persisted to the backend yet.
 * That means it might be missing any of the url fields
 *
 * It can be saved to backend once at least the `key` field is set, as that one
 * can be used to infer/compute all the other fields.
 *
 * - `crop_key` - repleace "original" with "crop" in `key` field
 * - `crop_url` - signed s3 url (by backend), for `crop_key`
 * - `original_image_url` - signed s3 url (by backend) for `key`
 */
export type PendingClassImage =
  Omit<AnnotationClassPayload['images'][0], PendingFields>
  & Partial<Pick<AnnotationClassPayload['images'][0], PendingFields>>

// eslint-disable-next-line camelcase
export type PendingClassImageWithUrlSet = PendingClassImage & { original_image_url: string }

export const isClassImage = (
  image: PendingClassImage
): image is AnnotationClassPayload['images'][0] =>
  !!image.key && !!image.crop_key

export type AnnotationClassData = {
  datasets: AnnotationClassPayload['datasets']
  description: string | null
  hotkey: null | string
  id: number | null
  images: PendingClassImage[]
  metadata: AnnotationClassMetadata
  name: string | null
}

export type DefaultAnnotationClassData = Partial<AnnotationClassData> & {
  selectedMainAnnotationType?: AnnotationTypeName
}

export type AnnotationClassValidationErrors = {
  annotationTypes?: string
  name?: string
  skeleton?: string
}

export const VIEWPORT_WIDTH: number = 150
export const VIEWPORT_HEIGHT: number = 150

export type CropInfo = {
  orientation: number
  points: [string, string, string, string]
  zoom: number
}

/**
 * VueCroppie is really archaic and requires us to get crop info via $ref,
 * so this type defines that ref
 */
export type Croppie = {
  /** Binds options to the underlying croppie library */
  bind: (options: Pick<CropInfo, 'points' | 'zoom'> & { url: string}) => void

  /** Retrieves current crop info */
  get: () => CropInfo
  result: (options?: Record<string, string>, callback?: Function) => void

  /** Sets scale/zoom level of cropper */
  setZoom: (value: number) => void
}
