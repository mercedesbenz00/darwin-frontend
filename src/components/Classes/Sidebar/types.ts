import { AnnotationTypeName } from '@/store/types'

export type AnnotationTypeFilterItemType = {
  id: string | number
  label: string
  icon?: string
  name: AnnotationTypeName
  color: string
  count: number | null
  [key: string]: any
}
