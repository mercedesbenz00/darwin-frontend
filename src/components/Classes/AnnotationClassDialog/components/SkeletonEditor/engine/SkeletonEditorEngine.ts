import { v4 as uuidv4 } from 'uuid'

import { SkeletonMetadata } from '@/store/types'

import SkeletonEditorRenderer from './SkeletonEditorRenderer'
import SkeletonEditorSerializer from './SkeletonEditorSerializer'
import { SkeletonNodeType, SkeletonEdgeType, NODE_SIZE } from './types'

export default class SkeletonEditorEngine {
  private repaintHandle: number | null = null
  private editing?: boolean
  private scale: number = 1
  private serializer: SkeletonEditorSerializer = new SkeletonEditorSerializer(this)
  private renderer: SkeletonEditorRenderer = new SkeletonEditorRenderer(this)
  private originalRawData: SkeletonMetadata | null = null

  public canvas!: HTMLCanvasElement
  public edges: SkeletonEdgeType[] = []
  public nodes: SkeletonNodeType[] = []
  public strokeColor: string = 'rgb(49, 245, 202)'
  public width: number = 0
  public height: number = 0

  private onChange: (data: SkeletonMetadata) => void

  constructor (onChange: (data: SkeletonMetadata) => void) {
    this.onChange = onChange
  }

  /**
   * Returns true if no edges, no nodes
   */
  public get isEmpty (): boolean {
    return this.edges.length === 0 && this.nodes.length === 0
  }

  public get hasMovingNode (): boolean {
    return !!this.movingNode
  }

  public get movingNode (): SkeletonNodeType | undefined {
    return this.nodes.find(node => node.isMoving)
  }

  public get hasDrawingEdge (): boolean {
    return !!this.drawingEdge
  }

  public get drawingEdge (): SkeletonEdgeType | undefined {
    return this.edges.find(edge => edge.isDrawing)
  }

  // The node is highlighted when the mouse is over its node's area
  public get highlightedNode (): SkeletonNodeType | undefined {
    return this.nodes.find(node => node.isHighlighted)
  }

  public setMainCanvas (canvas: HTMLCanvasElement): void {
    this.canvas = canvas
    this.resetDimensions()
    this.addListeners()
  }

  public getRawData (): SkeletonMetadata {
    return this.serializer.serialize()
  }

  public setRawData (rawData: SkeletonMetadata): void {
    this.originalRawData = rawData
    this.serializer.deserialize(rawData)
    this.requestRepaint()
  }

  public setNodes (nodes: SkeletonNodeType[]): void {
    this.nodes = nodes
  }

  public setEdges (edges: SkeletonEdgeType[]): void {
    this.edges = edges
  }

  private cancelAllMovingNodes (): void {
    const { nodes } = this
    nodes.forEach(node => { node.isMoving = false })
  }

  public setMovingNode (node: SkeletonNodeType): void {
    this.cancelAllMovingNodes()
    node.isMoving = true
  }

  public startDrawingEdgeFromNode (node: SkeletonNodeType): void {
    this.edges.push({
      internalId: uuidv4(),
      nodes: [
        node,
        {
          position: { x: node.position.x, y: node.position.y },
          isMoving: false,
          isHighlighted: false
        }
      ],
      isDrawing: true
    })
  }

  private completeDrawingEdge (): void {
    const { edges, highlightedNode, drawingEdge } = this
    if (!drawingEdge || !highlightedNode) { return }

    drawingEdge.isDrawing = false

    // If it is creating a new edge, create a new one.
    // If there is an existing edge between the same nodes, remove it.
    const existingEdge = edges.find(
      edge => edge.internalId !== drawingEdge.internalId &&
        edge.nodes.every(node => (
          node.internalId === drawingEdge.nodes[0].internalId ||
          node.internalId === highlightedNode.internalId
        ))
    )

    if (existingEdge) {
      this.removeEdge(existingEdge)
      this.removeEdge(drawingEdge)
    } else {
      drawingEdge.nodes.splice(1, 1, highlightedNode)
    }

    this.refreshHighlightedNodes()
    this.requestRepaint()
    this.onChange(this.getRawData())
  }

  private cancelDrawingEdge (): void {
    const { edges, drawingEdge } = this
    if (!drawingEdge) { return }

    const idx = edges.findIndex(edge => edge.internalId === drawingEdge.internalId)
    edges.splice(idx, 1)

    this.requestRepaint()
    this.onChange(this.getRawData())
  }

  private isPointOverNode (node: SkeletonNodeType, point: { x: number, y: number }): boolean {
    const { x, y } = node.position
    const { scale } = this
    return Math.abs(x - point.x) < NODE_SIZE * scale &&
      Math.abs(y - point.y) < NODE_SIZE * scale
  }

  private refreshHighlightedNodes (point?: { x: number, y: number }): void {
    this.nodes.forEach(node => {
      node.isHighlighted = point ? this.isPointOverNode(node, point) : false
    })
  }

  public addNewNodeFromPosition (position: { x: number, y: number }): void {
    const newNode = {
      internalId: uuidv4(),
      position,
      label: `${this.nodes.length + 1}`,
      isMoving: false,
      isHighlighted: false
    }
    this.nodes.push(newNode)
  }

  public removeNode (node: SkeletonNodeType): void {
    const { edges, nodes } = this
    // Removing related edges
    for (let i = edges.length - 1; i >= 0; i--) {
      const edge = edges[i]
      if (edge.nodes.includes(node)) {
        edges.splice(i, 1)
      }
    }

    const idx = nodes.findIndex(n => n.internalId === node.internalId)
    nodes.splice(idx, 1)

    this.requestRepaint()
    this.onChange(this.getRawData())
  }

  private removeEdge (edge: SkeletonEdgeType): void {
    const { edges } = this
    const idx = edges.findIndex(e => e.internalId === edge.internalId)
    edges.splice(idx, 1)
  }

  public setEditing (editing: boolean): void {
    this.editing = editing
  }

  public setStrokeColor (color: string): void {
    this.strokeColor = color === 'auto' ? 'rgb(49, 245, 202)' : color
  }

  public resetDimensions (): void {
    if (!this.canvas) { return }
    this.width = this.canvas.clientWidth
    this.height = this.canvas.clientHeight
    this.canvas.width = this.width
    this.canvas.height = this.height

    this.originalRawData && this.setRawData(this.originalRawData)
  }

  public setScale (scale: number): void {
    this.scale = scale
  }

  public reset (): void {
    this.removeListeners()
  }

  private addListeners (): void {
    if (!this.canvas) { return }
    if (!this.canvas.parentElement) { return }

    this.canvas.parentElement.addEventListener('mouseup', this.onMouseUp.bind(this))

    this.canvas.addEventListener('mousedown', this.onCanvasMouseDown.bind(this))
    this.canvas.addEventListener('mousemove', this.onCanvasMouseMove.bind(this))

    document.addEventListener('keydown', this.onKeydown.bind(this))
  }

  private removeListeners (): void {
    if (!this.canvas) { return }
    if (!this.canvas.parentElement) { return }

    this.canvas.parentElement.removeEventListener('mouseup', this.onMouseUp.bind(this))

    this.canvas.removeEventListener('mousedown', this.onCanvasMouseDown.bind(this))
    this.canvas.removeEventListener('mousemove', this.onCanvasMouseMove.bind(this))

    document.removeEventListener('keydown', this.onKeydown.bind(this))
  }

  onMouseUp (event: MouseEvent): void {
    if (!this.canvas.parentElement) { return }
    if (!this.canvas.parentElement.contains(event.target as HTMLElement)) { return }

    const node = this.movingNode
    if (!node) { return }

    this.cancelAllMovingNodes()
  }

  onCanvasMouseDown (event: MouseEvent): void {
    const { editing, highlightedNode, movingNode, drawingEdge } = this

    const shouldCompleteDrawingEdge = !!drawingEdge && !!highlightedNode
    const shouldCancelDrawingEdge = !!drawingEdge && !highlightedNode
    const isCreatingNewNode = !drawingEdge && !movingNode && !editing

    if (shouldCompleteDrawingEdge) { this.completeDrawingEdge() }
    if (shouldCancelDrawingEdge) { this.cancelDrawingEdge() }
    if (isCreatingNewNode) { this.addNewNodeFromPosition({ x: event.offsetX, y: event.offsetY }) }
  }

  onCanvasMouseMove (event: MouseEvent): void {
    const { movingNode, drawingEdge } = this
    if (drawingEdge) {
      drawingEdge.nodes[1].position.x = event.offsetX
      drawingEdge.nodes[1].position.y = event.offsetY

      this.refreshHighlightedNodes({ x: event.offsetX, y: event.offsetY })
      this.requestRepaint()
    } else if (movingNode) {
      movingNode.position.x = event.offsetX
      movingNode.position.y = event.offsetY
      this.requestRepaint()
    }
    this.onChange(this.getRawData())
  }

  onKeydown (event: KeyboardEvent): void {
    if (event.key === 'Escape') {
      event.stopPropagation()
      this.cancelDrawingEdge()
    }
  }

  public requestRepaint (): boolean {
    if (this.repaintHandle) { window.cancelAnimationFrame(this.repaintHandle) }

    this.repaintHandle = window.requestAnimationFrame(() => {
      this.renderer.render()
      this.repaintHandle = null
    })

    return true
  }
}
