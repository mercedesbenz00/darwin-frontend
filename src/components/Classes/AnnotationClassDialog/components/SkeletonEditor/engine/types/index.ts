export * from './SkeletonEdgeType'
export * from './SkeletonNodeType'
