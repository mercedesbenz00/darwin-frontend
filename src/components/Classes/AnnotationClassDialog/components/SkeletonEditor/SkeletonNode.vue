<template>
  <div
    class="skeleton-node"
    :class="{
      'skeleton-node--no-action': engine.hasDrawingEdge || engine.hasMovingNode,
      'skeleton-node--highlighted': node.isHighlighted
    }"
    :style="nodeStyle"
  >
    <skeleton-node-id
      class="skeleton-node__id"
      :label="node.label"
    />
    <div class="skeleton-node__actions">
      <div
        class="skeleton-node__main"
        :style="{ borderColor: engine.strokeColor }"
      />
      <skeleton-node-overlay
        class="skeleton-node__overlay"
        :engine="engine"
        :node="node"
        @change="$emit('change', node)"
      />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

import SkeletonNodeId from './SkeletonNodeId.vue'
import SkeletonNodeOverlay from './SkeletonNodeOverlay.vue'
import SkeletonEditorEngine from './engine/SkeletonEditorEngine'
import { SkeletonNodeType } from './engine/types'

@Component({
  name: 'skeleton-node',
  components: { SkeletonNodeId, SkeletonNodeOverlay }
})
export default class SkeletonNode extends Vue {
  @Prop({ required: true })
  engine!: SkeletonEditorEngine

  @Prop({ required: true })
  node!: SkeletonNodeType

  get nodeStyle (): { top: string, left: string } {
    const { x, y } = this.node.position
    return {
      top: `${y}px`,
      left: `${x}px`
    }
  }
}
</script>

<style lang="scss" scoped>
.skeleton-node {
  position: absolute;
  cursor: pointer;
}

.skeleton-node__actions {
  position: absolute;
  top: 0;
  left: 0;

  &:hover {
    .skeleton-node__main {
      display: none;
    }

    .skeleton-node__overlay {
      opacity: 1;
      transform: translateX(-50%) translateY(-50%);
      transition: all .3s;
      pointer-events: unset;
    }
  }
}

.skeleton-node--no-action {
  .skeleton-node__actions {
    pointer-events: none;
  }
}

.skeleton-node__main {
  position: absolute;
  top: 50%;
  left: 50%;
  width: 15px;
  height: 15px;
  background: white;
  border-radius: 50%;
  border-width: 1px;
  border-style: solid;
  border-color: $colorFeatherLight;
  transform: translateX(-50%) translateY(-50%);
  transition: all .2s;
}

.skeleton-node--highlighted {
  .skeleton-node__main {
    width: 20px;
    height: 20px;
  }
}

.skeleton-node__id {
  position: absolute;
  top: -30px;
  left: 50%;
  transform: translateX(-50%);
  pointer-events: none;
}

.skeleton-node__overlay {
  position: absolute;
  top: 50%;
  left: 50%;
  opacity: 0;
  transform: translateX(-50%) translateY(-50%);
  transition: all .3s;
  pointer-events: none;
}
</style>
