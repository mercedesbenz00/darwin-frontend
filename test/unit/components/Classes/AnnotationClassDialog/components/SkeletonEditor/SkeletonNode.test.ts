import { createLocalVue, shallowMount } from '@vue/test-utils'

import { buildSkeletonNodeType, buildSkeletonEdgeType } from 'test/unit/factories'

import SkeletonNode from '@/components/Classes/AnnotationClassDialog/components/SkeletonEditor/SkeletonNode.vue'
import SkeletonEditorEngine from '@/components/Classes/AnnotationClassDialog/components/SkeletonEditor/engine/SkeletonEditorEngine'
import { SkeletonNodeType } from '@/components/Classes/AnnotationClassDialog/components/SkeletonEditor/engine/types'

const localVue = createLocalVue()

let engine: SkeletonEditorEngine
let propsData: {
  engine: SkeletonEditorEngine
  node: SkeletonNodeType
}

beforeEach(() => {
  engine = new SkeletonEditorEngine(jest.fn())
  engine.setStrokeColor('#0000ff')
  propsData = {
    engine,
    node: buildSkeletonNodeType()
  }
})

it('matches snapshot', () => {
  const wrapper = shallowMount(SkeletonNode, { localVue, propsData })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot when drawing', () => {
  engine.setEdges([buildSkeletonEdgeType({ isDrawing: true })])
  const wrapper = shallowMount(SkeletonNode, { localVue, propsData })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot when node is highlighted', () => {
  propsData.node = buildSkeletonNodeType({ isHighlighted: true })
  const wrapper = shallowMount(SkeletonNode, { localVue, propsData })
  expect(wrapper).toMatchSnapshot()
})

it('should emit change when change is detected', async () => {
  const wrapper = shallowMount(SkeletonNode, { localVue, propsData })
  await wrapper.find('skeleton-node-overlay-stub').vm.$emit('change')
  expect(wrapper.emitted().change).toEqual([[propsData.node]])
})
