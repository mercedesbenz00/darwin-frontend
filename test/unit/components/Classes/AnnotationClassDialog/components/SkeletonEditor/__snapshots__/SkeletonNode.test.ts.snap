// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div class="skeleton-node" style="top: 0px; left: 0px;">
  <skeleton-node-id-stub label="Label" class="skeleton-node__id"></skeleton-node-id-stub>
  <div class="skeleton-node__actions">
    <div class="skeleton-node__main" style="border-color: #0000ff;"></div>
    <skeleton-node-overlay-stub engine="[object Object]" node="[object Object]" class="skeleton-node__overlay"></skeleton-node-overlay-stub>
  </div>
</div>
`;

exports[`matches snapshot when drawing 1`] = `
<div class="skeleton-node skeleton-node--no-action" style="top: 0px; left: 0px;">
  <skeleton-node-id-stub label="Label" class="skeleton-node__id"></skeleton-node-id-stub>
  <div class="skeleton-node__actions">
    <div class="skeleton-node__main" style="border-color: #0000ff;"></div>
    <skeleton-node-overlay-stub engine="[object Object]" node="[object Object]" class="skeleton-node__overlay"></skeleton-node-overlay-stub>
  </div>
</div>
`;

exports[`matches snapshot when node is highlighted 1`] = `
<div class="skeleton-node skeleton-node--highlighted" style="top: 0px; left: 0px;">
  <skeleton-node-id-stub label="Label" class="skeleton-node__id"></skeleton-node-id-stub>
  <div class="skeleton-node__actions">
    <div class="skeleton-node__main" style="border-color: #0000ff;"></div>
    <skeleton-node-overlay-stub engine="[object Object]" node="[object Object]" class="skeleton-node__overlay"></skeleton-node-overlay-stub>
  </div>
</div>
`;
